src/t_hash.c

```diff
@@ -929,8 +929,10 @@ int hashTypeSet(redisDb *db, robj *o, sds field, sds value, int flags) {
         }
         o->ptr = zl;
 
-        /* Check if the listpack needs to be converted to a hash table */
-        if (hashTypeLength(o, 0) > server.hash_max_listpack_entries)
+        /* Check if the listpack needs to be converted to a hash table.
+         * We do not subtract expired field here, if there is no new field inserted,
+         * we also skip recomputing the listpack element count, avoiding unnecessary overhead. */
+        if (!update && hashTypeLength(o, 0) > server.hash_max_listpack_entries)
```

Comment:
but we still might replace a small field with a large field.
in this case, does hash still need to be converted to ht?

Reply:
The size of the field has been validated here:
https://github.com/redis/redis/blob/ba88a7fbb6ec7cb7aeda61e1981146e1a21dee43/src/t_hash.c#L898-L902

The function `hashTypeLength` only checks the number of elements in the hash.